File: src/main/java/de/byjoker/myjfql/command/StructureCommand.java
package de.byjoker.myjfql.command;

import de.byjoker.myjfql.core.MyJFQL;
import de.byjoker.myjfql.database.*;
import de.byjoker.myjfql.server.session.Session;
import de.byjoker.myjfql.util.ResultType;
import org.jetbrains.annotations.NotNull;

import java.util.*;
import java.util.stream.Collectors;

@CommandHandler
public class StructureCommand extends Command {

    public StructureCommand() {
        super("structure", Arrays.asList("COMMAND", "ADD", "SET", "REMOVE", "OF", "MARK-PRIMARY", "PRIMARY-KEY"));
    }

    @Override
    public void execute(CommandSender sender, @NotNull Map<String, List<String>> args) {
        final DatabaseService databaseService = MyJFQL.getInstance().getDatabaseService();
        final Session session = sender.getSession();

        if (session == null) {
            sender.sendError("Session of this user is invalid!");
            return;
        }

        final Database database = session.getDatabase(databaseService);

        if (database == null) {
            sender.sendError("No database is in use for this user!");
            return;
        }

        if (args.containsKey("OF")) {
            final String name = formatString(args.get("OF"));

            if (name == null) {
                sender.sendError("Undefined table!");
                return;
            }

            if (!database.existsTable(name)) {
                sender.sendError("Table doesn't exists!");
                return;
            }


            if (!sender.allowed(database.getId(), DatabasePermissionLevel.READ)) {
                sender.sendForbidden();
                return;
            }

            final Table table = database.getTable(name);
            final Collection<String> structure = table.getStructure();
            final String primary = table.getPrimary();

            if (!args.containsKey("ADD") && !args.containsKey("REMOVE") && !args.containsKey("SET") && !args.containsKey("MARK-PRIMARY")) {
                if (args.containsKey("PRIMARY-KEY")) {
                    sender.sendResult(Collections.singletonList(new RelationalTableEntry().append("primary_key_name", primary)), Collections.singletonList("primary_key_name"), ResultType.RELATIONAL);
                    return;
                }

                sender.sendResult(structure.stream().map(s -> new RelationalTableEntry().append("field_name", s)).collect(Collectors.toList()), Collections.singletonList("field_name"), ResultType.RELATIONAL);
                return;
            }

            if (!sender.allowed(database.getId(), DatabasePermissionLevel.READ_WRITE)) {
                sender.sendForbidden();
                return;
            }

            if (args.containsKey("ADD")) {
                final String argument = formatString(args.get("ADD"));

                if (structure.contains(argument)) {
                    sender.sendError("Key already contains in structure of table!");
                    return;
                }

                structure.add(argument);
                sender.sendSuccess();

                table.setStructure(structure);
                database.saveTable(table);
                databaseService.saveDatabase(database);
                return;
            }

            if (args.containsKey("REMOVE")) {
                final String argument = formatString(args.get("REMOVE"));

                if (!structure.contains(argument)) {
                    sender.sendError("Key doesn't contains in structure of table!");
                    return;
                }

                if (primary.equals(argument)) {
                    sender.sendError("Primary key can't be removed!");
                    return;
                }

                structure.remove(argument);

                sender.sendSuccess();

                table.setStructure(structure);
                database.saveTable(table);
                databaseService.saveDatabase(database);
                return;
            }

            if (args.containsKey("MARK-PRIMARY")) {
                final String argument = formatString(args.get("MARK-PRIMARY"));

                if (!structure.contains(argument)) {
                    sender.sendError("Key doesn't contains in structure of table!");
                    return;
                }

                if (primary.equals(argument)) {
                    sender.sendError("Key is already the primary key!");
                    return;
                }

                table.setPrimary(argument);

                sender.sendSuccess();

                database.saveTable(table);
                databaseService.saveDatabase(database);
                return;
            }

            if (args.containsKey("SET")) {
                final List<String> newStructure = formatList(args.get("SET"));

                if (newStructure == null) {
                    sender.sendError("Undefined structure!");
                    return;
                }

                if (newStructure.size() == 0) {
                    sender.sendError("Structures size cant be 0!");
                    return;
                }

                if (structure.equals(newStructure)) {
                    sender.sendError("Structure is the same as before!");
                    return;
                }

                table.setStructure(newStructure);

                if (args.containsKey("PRIMARY-KEY")) {
                    final String key = formatString(args.get("PRIMARY-KEY"));

                    if (key == null) {
                        sender.sendError("Key can't be null!");
                        return;
                    }

                    if (!structure.contains(key)) {
                        sender.sendError("Unknown key!");
                        return;
                    }

                    table.setPrimary(key);
                } else {
                    table.setPrimary(newStructure.get(0));
                }


                sender.sendSuccess();

                database.saveTable(table);
                databaseService.saveDatabase(database);
                return;
            }

            return;
        }

        sender.sendSyntax();
    }
}

File: src/main/java/de/byjoker/myjfql/command/StructureCommand.kt
package de.byjoker.myjfql.command

import de.byjoker.myjfql.core.MyJFQL
import de.byjoker.myjfql.database.DatabasePermissionLevel
import de.byjoker.myjfql.database.DocumentCollection
import de.byjoker.myjfql.database.RelationalTable
import de.byjoker.myjfql.database.RelationalTableEntry
import de.byjoker.myjfql.util.ResultType

@CommandHandler
class StructureCommand :
    Command("structure", listOf("COMMAND", "OF", "SET", "ADD", "REMOVE", "MARK-PRIMARY", "PRIMARY-KEY")) {

    override fun execute(sender: CommandSender, args: Map<String, List<String>>) {
        val databaseService = MyJFQL.getInstance().databaseService
        val session = sender.session

        if (session == null) {
            sender.sendError("Session of this user is invalid!")
            return
        }

        val database = session.getDatabase(databaseService)

        if (database == null) {
            sender.sendError("No database is in use for this user!")
            return
        }

        if (!sender.allowed(database.id, DatabasePermissionLevel.READ_WRITE)) {
            sender.sendForbidden()
            return
        }

        if (!args.containsKey("OF")) {
            sender.sendSyntax()
            return
        }

        val tableIdenf = formatString(args["OF"])

        if (tableIdenf == null) {
            sender.sendError("Undefined table!")
            return
        }

        val table = database.getTable(tableIdenf)


        if (table == null) {
            sender.sendError("Table doesn't exists!")
            return
        }

        val structure = table.structure.toMutableList()
        val primary = table.primary

        if (args.size == 2) {
            val entries = structure.map { column ->
                val entry = RelationalTableEntry()
                entry.insert("name", column)

                if (column == primary) {
                    entry.insert("type", "PRIMARY_KEY")
                } else {
                    entry.insert("type", "COLUMN")
                }

                entry
            }

            sender.sendResult(entries, listOf("name", "type"), ResultType.RELATIONAL)
            return
        }

        if (args.containsKey("ADD")) {
            val columns = formatList(args["ADD"])

            if (columns == null) {
                sender.sendError("Undefined columns!")
                return
            }

            if (columns.any { column -> structure.contains(column) }) {
                sender.sendError("Columns already match table structure!")
                return
            }

            structure.addAll(columns)

            table.structure = structure
            database.saveTable(table)
            databaseService.saveDatabase(database)

            sender.sendSuccess();
            return
        }

        if (args.containsKey("REMOVE")) {
            val columns = formatList(args["REMOVE"])

            if (columns == null) {
                sender.sendError("Undefined columns!")
                return
            }

            if (table is RelationalTable && columns.any { column -> column == primary }) {
                sender.sendError("Primary key cannot be removed of the table structure!")
                return
            }

            if (columns.any { column -> !structure.contains(column) }) {
                sender.sendError("Columns doesn't match table structure!")
                return
            }

            structure.removeIf(columns::contains)

            table.structure = structure
            database.saveTable(table)
            databaseService.saveDatabase(database)

            sender.sendSuccess();
            return
        }

        if (args.containsKey("SET")) {
            val columns = formatList(args["STRUCTURE"])

            if (columns == null) {
                sender.sendError("Undefined table structure!")
                return
            }

            val primaryColumn: String? = when {
                args.containsKey("PRIMARY-KEY") -> formatString(args["PRIMARY-KEY"])
                else -> columns[0]
            }

            if (primaryColumn == null) {
                sender.sendError("Undefined primary key!")
                return
            }

            if (!columns.contains(primaryColumn)) {
                sender.sendError("Primary key does not match table structure!")
                return
            }

            table.structure = structure
            table.primary = primaryColumn
            database.saveTable(table)
            databaseService.saveDatabase(database)

            sender.sendSuccess()
            return
        }

        if (args.containsKey("MARK-PRIMARY")) {
            if (table is DocumentCollection) {
                sender.sendError("A document table has a fixed and not changeable primary key!")
                return
            }

            val column = formatString(args["MARK-PRIMARY"])

            if (column == null) {
                sender.sendError("Undefined primary key!")
                return
            }

            if (!structure.contains(column)) {
                sender.sendError("Primary key does not match table structure!")
                return
            }

            table.primary = column
            database.saveTable(table)
            databaseService.saveDatabase(database)

            sender.sendSuccess()
            return
        }


        sender.sendSyntax()
    }

}
